node/pkg/watchers/near/watcher.go

```diff
@@ -1,556 +1,309 @@
 		}
 	}
-
-	return t, last_block, nil
 }
 
-func (e *Watcher) inspectBody(logger *zap.Logger, block uint64, body gjson.Result) error {
-	logger.Info("inspectBody", zap.Uint64("block", block))
+func (e *Watcher) runTxProcessor(ctx context.Context) error {
+	logger := supervisor.Logger(ctx)
+	supervisor.Signal(ctx, supervisor.SignalHealthy)
 
-	result := body.Get("result.chunks.#.chunk_hash")
-	if !result.Exists() {
-		return nil
-	}
+	ticker := time.NewTicker(time.Microsecond * 100)
 
-	for _, name := range result.Array() {
-		chunk, err := e.getChunk(name.String())
-		if err != nil {
-			return err
-		}
+	for {
+		select {
+		case <-ctx.Done():
+			return ctx.Err()
 
-		txns := gjson.ParseBytes(chunk).Get("result.transactions")
-		if !txns.Exists() {
-			continue
-		}
-		for _, r := range txns.Array() {
-			hash := r.Get("hash")
-			receiver_id := r.Get("receiver_id")
-			if !hash.Exists() || !receiver_id.Exists() {
+		case <-ticker.C:
+			j, _, err := e.transactionProcessingQueue.PopFirstIfReady()
+			if err != nil {
 				continue
 			}
 
-			t, round, err := e.lastBlock(logger, hash.String(), receiver_id.String())
+			job := j.(transactionProcessingJob)
+
+			err = e.processTx(logger, ctx, &job)
 			if err != nil {
-				return err
-			}
-			if round != 0 {
-
-				if round <= e.final_round {
-					logger.Info("parseStatus direct", zap.Uint64("block.height", round), zap.Uint64("e.final_round", e.final_round))
-					err := e.parseStatus(logger, t, hash.String())
-					if err != nil {
-						return err
-					}
+				// transaction processing unsuccessful. Retry if retry_counter not exceeded.
+
+				if job.retryCounter < txProcRetry {
+					// Warn and retry with exponential backoff
+					logger.Warn(
+						"near.processTx",
+						zap.String("log_msg_type", "tx_processing_retry"),
+						zap.String("tx_hash", job.txHash),
+						zap.String("error", err.Error()),
+					)
+					job.retryCounter++
+					job.delay *= 2
+					e.transactionProcessingQueue.Schedule(job, time.Now().Add(job.delay))
 				} else {
-					logger.Info("pushing pending",
-						zap.Uint64("block.height", round),
-						zap.Uint64("e.final_round", e.final_round),
+					// Error and do not retry
+					logger.Error(
+						"near.processTx",
+						zap.String("log_msg_type", "tx_processing_retries_exceeded"),
+						zap.String("tx_hash", job.txHash),
+						zap.String("error", err.Error()),
 					)
-					key := pendingKey{
-						hash: hash.String(),
-					}
-
-					e.pendingMu.Lock()
-					e.pending[key] = &pendingMessage{
-						height: round,
-					}
-					e.pendingMu.Unlock()
+					p2p.DefaultRegistry.AddErrorCount(vaa.ChainIDNear, 1)
 				}
 			}
-		}
 
+			if job.hasWormholeMsg {
+				// report how long it took to process this transaction
+				e.eventChanTxProcessedDuration <- time.Since(job.creationTime)
+			}
+
+			// tell everyone about successful processing
+			e.eventChanBlockProcessedHeight <- job.wormholeMsgBlockHeight
```

Comment:
Since transactions can be processed out of order, should this be where the block height gets written?

Reply:
Are you asking if this is the correct place to write in there or are you asking to do something additional here? 
In metrics.go, we have an additional check so we only report the max. of heights.